### PR TITLE
params, crypto, core, core/vm: homestead consensus protocol changes

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**
**WIP!**
One test failing with tests repo Homestead branch: **bcWalletTest wallet2outOf3txs**
**Exact homestead block number is yet to be configured once we have community consensus.**

https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.mediawiki

tests branch: https://github.com/ethereum/tests/tree/Homestead

* change gas cost for contract creating txs
* invalidate signature with s value greater than secp256k1 N / 2
* OOG contract creation if not enough gas to store code
* new difficulty adjustment algorithm

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Dec 1 11:34:51 UTC 2015_

File: common/big.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ var (
 	Big0     = big.NewInt(0)
 	BigTrue  = Big1
 	BigFalse = Big0
+	Big10    = big.NewInt(10)
```
Comment (obscuren) [inline common/big.go]: Please remove this one, it's only being used once.

And perhaps this is more a general notion I don't want to use these silly `common` obscure and dangerous variables anymore. Put them in the package they are used instead (even when dup).

File: core/vm/instructions.go [hunk 2]
```diff
@@ -529,10 +529,11 @@ func opCreate(instr instruction, pc *uint64, env Environment, contract *Contract
 		dataGas.Mul(dataGas, params.CreateDataGas)
 		if contract.UseGas(dataGas) {
 			env.Db().SetCode(addr, ret)
+			stack.push(addr.Big())
+		} else if params.IsHomestead(env.BlockNumber()) {
+			env.Db().Delete(addr)
+			stack.push(new(big.Int))
 		}
```
Comment (holiman) [inline core/vm/instructions.go]: I believe you're missing a case here; `contract.UseGas(dataGas) == false` and not homestead. Before this commit; that case still pushed address to stack. As I read it right now, you will fail to push the address for pre-homestead contract creation failures. 
I think it should be something like this:

		stack.push(addr.Big())
	}
	else 
	{
		if params.IsHomestead(env.BlockNumber()) {
			env.Db().Delete(addr)
			stack.push(new(big.Int))
		}else{			
			stack.push(addr.Big)
		}
	}

Comment (Gustav-Simonsson) [inline core/vm/instructions.go]: Correct. A left-over from a previous version which had the "everything else" case outside the if clause. Another edge case we need a `StateTests` for.

File: core/block_validator.go [hunk 3]
```diff
@@ -111,6 +117,7 @@ func (v *BlockValidator) ValidateState(block, parent *types.Block, statedb *stat
 	// For valid blocks this should always validate to true.
 	rbloom := types.CreateBloom(receipts)
 	if rbloom != header.Bloom {
+		//fmt.Printf("FUNKY: ValidateState: block number: %v\n", block.Number())
```
Comment (obscuren) [inline core/block_validator.go]: Remove

File: core/block_validator.go [hunk 4]
```diff
@@ -241,3 +248,127 @@ func ValidateHeader(pow pow.PoW, header *types.Header, parent *types.Header, che
...
+		x.Set(bigMinus99)
+	}
+
+	// (parent_diff + parent_diff // 2048 * max(1 - (block_timestamp - parent_timestamp) // 10, -99))
+	y.Div(parentDiff, params.DifficultyBoundDivisor)
+	x.Mul(y, x)
+	x.Add(parentDiff, x)
+
+	// minimum difficulty can ever be (before exponential factor)
+	if x.Cmp(params.MinimumDifficulty) < 0 {
```
Comment (obscuren) [inline core/block_validator.go]: FYI there's a `common.BigMax`

File: core/block_validator.go [hunk 5]
```diff
@@ -241,3 +248,127 @@ func ValidateHeader(pow pow.PoW, header *types.Header, parent *types.Header, che
...
+	x := new(big.Int)
+	y := new(big.Int)
+
+	// 1 - (block_timestamp -parent_timestamp) // 10
+	x.Sub(bigTime, bigParentTime)
+	x.Div(x, big10)
+	x.Sub(common.Big1, x)
+
+	// max(1 - (block_timestamp - parent_timestamp) // 10, -99)))
+	if x.Cmp(bigMinus99) < 0 {
```
Comment (obscuren) [inline core/block_validator.go]: FYI there's a `common.BigMax`

File: core/block_validator.go (hunk 4 above)
Comment (obscuren) [inline core/block_validator.go]: Same as above

File: core/execution.go [hunk 6]
```diff
@@ -94,15 +101,62 @@ func exec(env vm.Environment, caller vm.ContractRef, address, codeAddr *common.A
 		}
 	}
 	env.Transfer(from, to, value)
+	snapshotPostTransfer := env.MakeSnapshot()
```
Comment (obscuren) [inline core/execution.go]: This (and the code below) should only happen for after the homestead block. Did I miss a check somewhere?

File: core/state_transition.go [hunk 7]
```diff
@@ -214,6 +235,18 @@ func (self *StateTransition) transitionDb() (ret []byte, usedGas *big.Int, err e
 			if err := self.useGas(dataGas); err == nil {
 				self.state.SetCode(addr, ret)
 			} else {
+				if homestead {
+					// rollback all contract creation changes except for
+					// sender's incremented account nonce and gas usage
+					// TODO: fucking gas hack... verify potential DoS vuln
```
Comment (obscuren) [inline core/state_transition.go]: Could you explain the vuln?

File: core/transaction_pool.go [hunk 8]
```diff
@@ -182,18 +183,26 @@ func (pool *TxPool) validateTx(tx *types.Transaction) error {
...
+	if err != nil {
+		return err
+	}
+
+	homestead := params.IsHomestead(GetHeadBlockNum(currentState.GetDB()))
+
 	// Validate the transaction sender and it's sig. Throw
 	// if the from fields is invalid.
-	if from, err = tx.From(); err != nil {
+	if homestead {
```
Comment (obscuren) [inline core/transaction_pool.go]: Is there a reason why this (and a few others) aren't in `From` directly?

File: core/vm/jump_table.go [hunk 9]
```diff
@@ -62,9 +62,10 @@ func init() {
 	jumpTable[PC] = jumpPtr{nil, true}
 	jumpTable[MSIZE] = jumpPtr{opMsize, true}
 	jumpTable[GAS] = jumpPtr{opGas, true}
-	jumpTable[CREATE] = jumpPtr{opCreate, true}
+	jumpTable[CREATE] = jumpPtr{nil, true}
 	jumpTable[CALL] = jumpPtr{opCall, true}
 	jumpTable[CALLCODE] = jumpPtr{opCallCode, true}
+	jumpTable[DELEGATECALL] = jumpPtr{opDelegateCall, true}
```
Comment (obscuren) [inline core/vm/jump_table.go]: This should only ever be here IF we're after the homestead block.

File: core/vm/vm.go [hunk 10]
```diff
@@ -201,6 +199,35 @@ func (self *Vm) Run(contract *Contract, input []byte) (ret []byte, err error) {
 
 						continue
 					}
+				case CREATE:
```
Comment (obscuren) [inline core/vm/vm.go]: Why did this get added to the switch statement? Is it because of `params.IsHomestead`?

Comment (obscuren) [inline core/vm/vm.go]: I see now. It's because of the error. Why not change the definition of the call to return an error instead.

**Comment by obscuren** [comment]
I can't find any changes made to the JIT. The JIT, being an experimental option, should still be in agreement.

**Comment by obscuren** [comment]
See https://github.com/ethereum/go-ethereum/pull/2116

**Pull request closed without merging**